### PR TITLE
Hw7: Add message broker

### PR DESCRIPTION
**Pull request by AnnFedorenko**

### CODE REVIEW

File: log-consumer/cmd/log-consumer/main.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+func failOnError(err error, msg string) {
+	if err != nil {
+		log.Fatalf("%s: %s", msg, err)
+	}
+}
+
+func main() {
+	err := godotenv.Load(".env")
+	failOnError(err, "Failed to load env")
+	rmqClient, _ := rmq.NewRMQClient(rmq.RMQConfig{
```
Comment (platinum139) [inline log-consumer/cmd/log-consumer/main.go:21]: You have unhandled error here. If an error happens while creating RabbitMQ Client, the client will be nil. But your program goes on and then you try to call some methods of the client. Your program will crash, which is not good.

File: log-consumer/cmd/log-consumer/main.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	logLevel := os.Getenv("LOG_LEVEL")
+
+	err = rmqClient.QueueDeclare(logLevel)
+	failOnError(err, "Failed to declare a queue")
+	defer rmqClient.CloseConnection()
+
+	msgs, err := rmqClient.Consume(logLevel)
+	failOnError(err, "Failed to register a consumer")
+	defer rmqClient.CloseConnection()
```
Comment (platinum139) [inline log-consumer/cmd/log-consumer/main.go:37]: You don't have to call defer 2 times. It should be called one time after successful rabbit mq client creation.

File: rate-app/cmd/rate-api/app.go [hunk 3]
```diff
@@ -23,22 +25,26 @@ var RateCreators = map[string]rate.RateFactory{
 }
 
 func Run() {
+	logCfg := InitLogConfig(config.Cfg)
+	rmqCl, _ := logmb.NewRMQClient(logCfg)
```
Comment (platinum139) [inline rate-app/cmd/rate-api/app.go:29]: Unhandled error here, same as in log-consumer.

Comment (platinum139) [reply rate-app/cmd/rate-api/app.go:29]: Also there is no closing connection for the client.

**Review by platinum139 — CHANGES_REQUESTED** [review]
In general, very good job!

**Review by platinum139 — APPROVED** [review]

File: log-consumer/cmd/log-consumer/main.go [hunk 4]
```diff
@@ -18,23 +18,23 @@ func failOnError(err error, msg string) {
...
-	rmqClient, _ := rmq.NewRMQClient(rmq.RMQConfig{
+	rmqClient, err := rmq.NewRMQClient(rmq.RMQConfig{
 		Host:     os.Getenv("RMQ_HOST"),
 		Port:     os.Getenv("RMQ_PORT"),
 		User:     os.Getenv("RMQ_USER"),
 		Password: os.Getenv("RMQ_PASSWORD"),
 		Exchange: os.Getenv("LOG_EXCHANGE"),
 	})
+	failOnError(err, "Failed to declare rabbitMQ consumer")
+	defer rmqClient.CloseConnection()
```
Comment (kerrrusha) [inline log-consumer/cmd/log-consumer/main.go:29]: Unhandled error
func (c *RMQClient) CloseConnection() error

File: rate-app/cmd/rate-api/app.go [hunk 5]
```diff
@@ -23,22 +25,30 @@ var RateCreators = map[string]rate.RateFactory{
...
-	rateServ := InitRateProvider()
+	rateServ := InitRateProvider(rmqLogger)
 
 	emailRepo := repository.NewEmailRepository(config.Cfg.EmailStorage)
 	emailServ := service.NewEmailService(emailRepo)
 
 	emailSendServ := service.NewEmailSendService(emailServ, rateServ, mailclient.NewSMTPClient(config.Cfg))
 
-	handler := handler.InitHandler(rateServ, emailServ, emailSendServ)
+	handler := handler.InitHandler(rateServ, emailServ, emailSendServ, rmqLogger)
```
Comment (kerrrusha) [inline rate-app/cmd/rate-api/app.go:45]: Variable 'handler' collides with imported package name

**Review by kerrrusha — COMMENTED** [review]
well done but can be better

**Review by kerrrusha — APPROVED** [review]
great job

**Pull request merged**